File: examples/python/client_kv_operations.py
#!/usr/bin/env python3
"""KV Operations Examples - Using ekodb_client Python library

This example demonstrates key-value operations using the Python client library.
Compare with kv_operations.py to see the difference!
"""

import asyncio
import os
from pathlib import Path

from dotenv import load_dotenv

# Import the ekodb_client (Python bindings for Rust client)
from ekodb_client import Client

# Load environment variables
env_path = Path(__file__).parent.parent / ".env"
load_dotenv(env_path)

BASE_URL = os.getenv("API_BASE_URL", "http://localhost:8080")
API_KEY = os.getenv("API_BASE_KEY", "a-test-api-key-from-ekodb")
SESSION_KEY = "kv_operations:py:session:user123"
PRODUCT_KEYS = [f"kv_operations:py:cache:product:{index}" for index in range(1, 4)]


async def run_examples(client):
    """Run KV operations examples using ekodb_client"""
    # Create ekoDB client
    print("✓ Client created")

    try:
        # Example 1: Set a key-value pair
        print("\n=== KV Set ===")
        await client.kv_set(SESSION_KEY, {"userId": 123, "username": "john_doe"})
        print("✓ Set key: session:user123")

        # Example 2: Get a value
        print("\n=== KV Get ===")
        result = await client.kv_get(SESSION_KEY)
        print(f"Retrieved value: {result}")

        # Example 3: Batch set multiple keys
        print("\n=== KV Batch Set ===")
        batch_entries = [
            {"key": PRODUCT_KEYS[0], "value": {"name": "Product 1", "price": 29.99}},
            {"key": PRODUCT_KEYS[1], "value": {"name": "Product 2", "price": 39.99}},
            {"key": PRODUCT_KEYS[2], "value": {"name": "Product 3", "price": 49.99}},
        ]
        set_results = await client.kv_batch_set(batch_entries)
        print(f"✓ Batch set {len(set_results)} keys")
        for key, was_set in set_results:
            print(f"  {key}: {'success' if was_set else 'failed'}")

        # Example 4: Batch get multiple keys
        print("\n=== KV Batch Get ===")
        keys = PRODUCT_KEYS
        batch_values = await client.kv_batch_get(keys)
        print(f"✓ Batch retrieved {len(batch_values)} values")
        for i, val in enumerate(batch_values):
            print(f"  {keys[i]}: {val}")

        # Example 5: Check if key exists
        print("\n=== KV Exists ===")
        exists = await client.kv_exists(SESSION_KEY)
        print(f"Key exists: {exists}")

        # Example 6: Find keys with pattern
        print("\n=== KV Find (Pattern Query) ===")
        cache_results = await client.kv_find(
            pattern="kv_operations:py:cache:product:.*"
        )
        print(f"Found {len(cache_results)} keys matching 'cache:product:.*'")

        # Example 7: Query all keys
        print("\n=== KV Query (Alias for Find) ===")
        all_results = await client.kv_query()
        print(f"Total keys in store: {len(all_results)}")

        # Example 8: Delete a key
        print("\n=== KV Delete ===")
        await client.kv_delete(SESSION_KEY)
        print("✓ Deleted key: session:user123")

        # Verify deletion with kv_exists
        exists_after = await client.kv_exists(SESSION_KEY)
        print(f"✓ Verified: Key exists after delete: {exists_after}")

        # Example 9: Batch delete multiple keys
        print("\n=== KV Batch Delete ===")
        delete_results = await client.kv_batch_delete(keys)
        print(f"✓ Batch deleted {len(delete_results)} keys")
        for key, was_deleted in delete_results:
            print(f"  {key}: {'deleted' if was_deleted else 'not found'}")

    except Exception as e:
        print(f"Error: {e}")
        raise


async def kv_operations_examples():
    client = Client.new(BASE_URL, API_KEY)
    keys = [SESSION_KEY, *PRODUCT_KEYS]
    for key in keys:
        if await client.kv_exists(key):
            await client.kv_delete(key)

    operation_error = None
    try:
        await run_examples(client)
    except BaseException as error:  # noqa: BLE001 - cleanup must run on cancellation
        operation_error = error
    cleanup_errors = []
    for key in keys:
        try:
            if await client.kv_exists(key):
                await client.kv_delete(key)
        except Exception as error:  # noqa: BLE001 - attempt every cleanup
            cleanup_errors.append(f"KV key {key}: {error}")
    if operation_error is not None:
        if cleanup_errors:
            operation_error.add_note(
                "cleanup also failed: " + "; ".join(cleanup_errors)
            )
        raise operation_error
    if cleanup_errors:
        raise RuntimeError("Cleanup failed: " + "; ".join(cleanup_errors))

    print("\n✓ All KV operations completed successfully")


if __name__ == "__main__":
    asyncio.run(kv_operations_examples())

File: examples/go/client_kv_operations.go
// Key-Value Operations - Using ekodb-client-go library
//
// This example demonstrates KV operations using the Go client library.
// Compare with kv_operations.go to see the difference!

package main

import (
	"errors"
	"fmt"
	"log"
	"os"

	ekodb "github.com/ekoDB/ekodb-client-go"
	"github.com/joho/godotenv"
)

const clientSessionKey = "session:user123:go"

var clientProductKeys = []string{"cache:product:1:go", "cache:product:2:go", "cache:product:3:go"}

func run() (runErr error) {
	// Load environment variables
	if err := godotenv.Load(); err != nil {
		log.Println("No .env file found")
	}

	baseURL := os.Getenv("API_BASE_URL")
	if baseURL == "" {
		baseURL = "http://localhost:8080"
	}

	apiKey := os.Getenv("API_BASE_KEY")
	if apiKey == "" {
		apiKey = "a-test-api-key-from-ekodb"
	}

	// Create ekoDB client
	client, err := ekodb.NewClient(baseURL, apiKey)
	if err != nil {
		return err
	}
	fmt.Println("✓ Client created")
	ownedKeys := make(map[string]bool)
	defer func() {
		for key := range ownedKeys {
			if err := client.KVDelete(key); err != nil {
				runErr = errors.Join(runErr, fmt.Errorf("cleanup KV key %s: %w", key, err))
			}
		}
	}()

	// Example 1: Set a key
	fmt.Println("\n=== KV Set ===")
	sessionData := map[string]interface{}{
		"userId":   123,
		"username": "john_doe",
	}
	if err := client.KVSet(clientSessionKey, sessionData); err != nil {
		return err
	}
	ownedKeys[clientSessionKey] = true
	fmt.Printf("✓ Set key: %s\n", clientSessionKey)

	// Example 2: Get a key
	fmt.Println("\n=== KV Get ===")
	value, err := client.KVGet(clientSessionKey)
	if err != nil {
		return err
	}
	fmt.Printf("Retrieved value: %v\n", value)

	// Example 3: Batch set multiple keys
	fmt.Println("\n=== KV Batch Set ===")
	batchEntries := []map[string]interface{}{
		{"key": clientProductKeys[0], "value": map[string]interface{}{"name": "Product 1", "price": 29.99}},
		{"key": clientProductKeys[1], "value": map[string]interface{}{"name": "Product 2", "price": 39.99}},
		{"key": clientProductKeys[2], "value": map[string]interface{}{"name": "Product 3", "price": 49.99}},
	}
	setResults, err := client.KVBatchSet(batchEntries)
	if err != nil {
		return err
	}
	for _, key := range clientProductKeys {
		ownedKeys[key] = true
	}
	fmt.Printf("✓ Batch set %d keys\n", len(setResults))
	for _, result := range setResults {
		key := result[0].(string)
		wasSet := result[1].(bool)
		status := "failed"
		if wasSet {
			status = "success"
		}
		fmt.Printf("  %s: %s\n", key, status)
	}

	// Example 4: Batch get multiple keys
	fmt.Println("\n=== KV Batch Get ===")
	keys := clientProductKeys
	batchValues, err := client.KVBatchGet(keys)
	if err != nil {
		return err
	}
	fmt.Printf("✓ Batch retrieved %d values\n", len(batchValues))
	for i, val := range batchValues {
		fmt.Printf("  %s: %v\n", keys[i], val)
	}

	// Example 5: Check if key exists
	fmt.Println("\n=== KV Exists ===")
	exists, err := client.KVExists(clientSessionKey)
	if err != nil {
		return err
	}
	fmt.Printf("Key exists: %v\n", exists)

	// Example 6: Find keys with pattern
	fmt.Println("\n=== KV Find (Pattern Query) ===")
	cacheResults, err := client.KVFind("cache:product:.*:go", false)
	if err != nil {
		return err
	}
	fmt.Printf("Found %d keys matching 'cache:product:.*:go'\n", len(cacheResults))

	// Example 7: Query all keys
	fmt.Println("\n=== KV Query (Alias for Find) ===")
	allResults, err := client.KVQuery("", false)
	if err != nil {
		return err
	}
	fmt.Printf("Total keys in store: %d\n", len(allResults))

	// Example 8: Delete a key
	fmt.Println("\n=== KV Delete ===")
	if err := client.KVDelete(clientSessionKey); err != nil {
		return err
	}
	delete(ownedKeys, clientSessionKey)
	fmt.Printf("✓ Deleted key: %s\n", clientSessionKey)

	// Verify deletion with KVExists
	existsAfter, err := client.KVExists(clientSessionKey)
	if err != nil {
		return err
	}
	fmt.Printf("✓ Verified: Key exists after delete: %v\n", existsAfter)

	// Example 9: Batch delete multiple keys
	fmt.Println("\n=== KV Batch Delete ===")
	deleteResults, err := client.KVBatchDelete(keys)
	if err != nil {
		return err
	}
	for _, key := range keys {
		delete(ownedKeys, key)
	}
	fmt.Printf("✓ Batch deleted %d keys\n", len(deleteResults))
	for _, result := range deleteResults {
		key := result[0].(string)
		wasDeleted := result[1].(bool)
		status := "not found"
		if wasDeleted {
			status = "deleted"
		}
		fmt.Printf("  %s: %s\n", key, status)
	}

	fmt.Println("\n✓ All KV operations completed successfully")
	return nil
}

func main() {
	if err := run(); err != nil {
		log.Fatal(err)
	}
}
